README.md
# Open COVID-19 Dataset

This repository contains datasets of daily time-series data related to COVID-19 for 30+ countries
around the world. For most countries, the data is at the spatial resolution of states/provinces,
although for US, UK, NL and CO, it is at the finer resolution of county/municipality. All regions
are assigned a unique key, which resolves discrepancies between ISO / NUTS / FIPS codes, etc.

There are multiple types of data:
* Outcome data `Y(i,t)`, such as cases, deaths, tests, for regions i and time t
* Static covariate data `X(i)`, such as population size, GDP, latitude/ longitude
* Dynamic covariate data `X(i,t)`, such as mobility, weather
* Dynamic interventional data `A(i,t)`, such as government lockdowns

The data is drawn from multiple sources, as listed [below](#sources-of-data), and stored in separate
csv / json files, which can be easily merged due to the use of consistent geographic (and temporal)
keys.

**NOTE: The URL endpoint for the data has changed, please see the updated links below**

| Table | Keys<sup>1</sup> | Content | URL | Source<sup>2</sup> |
| ----- | ---------------- | ------- | --- | ------------------ |
| [Main](#main) | `[key][date]` | Flat table with records from all other tables joined by `key` and `date` | [main.csv](https://storage.googleapis.com/covid19-open-data/v2/main.csv), [main.json](https://storage.googleapis.com/covid19-open-data/v2/main.json) | All tables below |
| [Index](#index) | `[key]` | Various names and codes, useful for joining with other datasets | [index.csv](https://storage.googleapis.com/covid19-open-data/v2/index.csv), [index.json](https://storage.googleapis.com/covid19-open-data/v2/index.json) | Wikidata, DataCommons |
| [Demographics](#demographics) | `[key]` | Various (current<sup>3</sup>) population statistics | [demographics.csv](https://storage.googleapis.com/covid19-open-data/v2/demographics.csv), [demographics.json](https://storage.googleapis.com/covid19-open-data/v2/demographics.json) | Wikidata, DataCommons |
| [Economy](#economy) | `[key]` | Various (current<sup>3</sup>) economic indicators | [economy.csv](https://storage.googleapis.com/covid19-open-data/v2/economy.csv), [economy.json](https://storage.googleapis.com/covid19-open-data/v2/economy.json) | Wikidata, DataCommons |
| [Epidemiology](#epidemiology) | `[key][date]` | COVID-19 cases, deaths, recoveries and tests | [epidemiology.csv](https://storage.googleapis.com/covid19-open-data/v2/epidemiology.csv), [epidemiology.json](https://storage.googleapis.com/covid19-open-data/v2/epidemiology.json) | Various<sup>2</sup> |
| [Geography](#geography) | `[key]` | Geographical information about the region | [geography.csv](https://storage.googleapis.com/covid19-open-data/v2/geography.csv), [geography.json](https://storage.googleapis.com/covid19-open-data/v2/geography.json) | Wikidata |
| [Health](#health) | `[key]` | Health indicators for the region | [health.csv](https://storage.googleapis.com/covid19-open-data/v2/health.csv), [health.json](https://storage.googleapis.com/covid19-open-data/v2/geography.json) | Wikidata, WorldBank |
| [Hospitalizations](#hospitalizations) | `[key][date]` | Information related to patients of COVID-19 and hospitals |  [hospitalizations.csv](https://storage.googleapis.com/covid19-open-data/v2/hospitalizations.csv), [hospitalizations.json](https://storage.googleapis.com/covid19-open-data/v2/hospitalization.json) | Various<sup>2</sup> |
| [Mobility](#mobility) | `[key][date]` | Various metrics related to the movement of people | [mobility.csv](https://storage.googleapis.com/covid19-open-data/v2/mobility.csv), [mobility.json](https://storage.googleapis.com/covid19-open-data/v2/mobility.json) | Google, Apple |
| [Oxford Government Response](#oxford-government-response) | `[key][date]` | Government interventions and their relative stringency | [oxford-government-response.csv](https://storage.googleapis.com/covid19-open-data/v2/oxford-government-response.csv), [oxford-government-response.json](https://storage.googleapis.com/covid19-open-data/v2/oxford-government-response.json) | University of Oxford |
| [Weather](#weather) | `[key][date]` | Dated meteorological information for each region | [weather.csv](https://storage.googleapis.com/covid19-open-data/v2/weather.csv), [weather.json](https://storage.googleapis.com/covid19-open-data/v2/weather.json) | NOAA |
| [WorldBank](#worldbank) | `[key]` | Latest record for each indicator from WorldBank for all reporting countries | [worldbank.csv](https://storage.googleapis.com/covid19-open-data/v2/worldbank.csv), [worldbank.json](https://storage.googleapis.com/covid19-open-data/v2/worldbank.json) | WorldBank |
| [WorldPop](#worldpop) | `[key]` | Demographics data extracted from WorldPop | [worldpop.csv](https://storage.googleapis.com/covid19-open-data/v2/worldpop.csv), [worldpop.json](https://storage.googleapis.com/covid19-open-data/v2/worldpop.json) | WorldPop |
| [By Age](#by-age) | `[key][date]` | Epidemiology and hospitalizations data stratified by age | [by-age.csv](https://storage.googleapis.com/covid19-open-data/v2/by-age.csv), [by-age.json](https://storage.googleapis.com/covid19-open-data/v2/by-age.json) | Various<sup>2</sup> |
| [By Sex](#by-sex) | `[key][date]` | Epidemiology and hospitalizations data stratified by sex | [by-sex.csv](https://storage.googleapis.com/covid19-open-data/v2/by-sex.csv), [by-sex.json](https://storage.googleapis.com/covid19-open-data/v2/by-sex.json) | Various<sup>2</sup> |

<sup>1</sup> `key` is a unique string for the specific geographical region built from a combination
of codes such as `ISO 3166`, `NUTS`, `FIPS` and other local equivalents.\
<sup>2</sup> Refer to the [data sources](#sources-of-data) for specifics about each data source and
the associated terms of use.\
<sup>3</sup> Datasets without a `date` column contain the most recently reported information for
each datapoint to date.

For more information about how to use these files see the section about
[using the data](#use-the-data), and for more details about each dataset see the section about
[understanding the data](#understand-the-data).



## Why another dataset?

There are many other public COVID-19 datasets. However, we believe this dataset is unique in the way
that it merges multiple global sources, at a fine spatial resolution, using a consistent set of
region keys. We hope this will make it easier for researchers to use. We are also very transparent
about the [data sources](#sources-of-data), and the [code](src/README.md) for ingesting and merging
the data is easy to understand and modify.



## Explore the data

|     |     |
| --- | --- |
| A simple visualization tool was built to explore the Open COVID-19 datasets, the [Open COVID-19 Explorer][12]: [![](https://github.com/open-covid-19/explorer/raw/master/screenshots/explorer.png)][12] | If you want to see [interactive charts with a unique UX][14], don't miss what [@Mahks](https://github.com/Mahks) built using the Open COVID-19 dataset: [![](https://i.imgur.com/cIODOtp.png)][14] |
| You can also check out the great work of [@quixote79](https://github.com/quixote79), [a MapBox-powered interactive map site][13]: [![](https://i.imgur.com/nFwxJId.png)][13] | Experience [clean, clear graphs with smooth animations][15] thanks to the work of [@jmullo](https://github.com/jmullo): [![](https://i.imgur.com/xdCzsUO.png)][15] |
| Become an armchair epidemiologist with the [COVID-19 timeline simulation tool][19] built by [@LeviticusMB](https://github.com/LeviticusMB): [![](https://i.imgur.com/4iWaP7E.png)][19] | Whether you want an interactive map, compare stats or look at charts, [@saadmas](https://github.com/saadmas) has you covered with a [COVID-19 Daily Tracking site][20]: [![](https://i.imgur.com/rAJvLSI.png)][20] |
| Compare per-million data at [Omnimodel][21] thanks to [@OmarJay1](https://github.com/OmarJay1): [![](https://i.imgur.com/RG7ZKXp.png)][21] | Look at responsive, comprehensive charts thanks to the work of [@davidjohnstone](https://github.com/davidjohnstone): [![](https://i.imgur.com/ZbfMKvu.png)](https://www.cyclinganalytics.com/covid19) |



## Use the data

The data is available as CSV and JSON files, which are published in Github Pages so they can be
served directly to Javascript applications without the need of a proxy to set the correct headers
for CORS and content type. Even if you only want the CSV files, using the URL served by Github Pages
is preferred in order to avoid caching issues and potential, future breaking changes.

For the purpose of making the data as easy to use as possible, there is a [main](#main) table
which contains the columns of all other tables joined by `key` and `date`. However,
performance-wise, it may be better to download the data separately and join the tables locally.

Each region has its own version of the main table, so you can pull all the data for a specific
region using a single endpoint, the URL for each region is:
* Data for `key` in CSV format: `https://storage.googleapis.com/covid19-open-data/v2/${key}/main.csv`
* Data for `key` in JSON format: `https://storage.googleapis.com/covid19-open-data/v2/${key}/main.json`

Each table has a full version as well as subsets with only the last 30, 14, 7 and 1 days of data.
The full version is accessible at the URL described [in the table above](#open-covid-19-dataset).
The subsets can be found by appending the number of days to the path. For example, the subsets of
the main table are available at the following locations:
* Full version: https://storage.googleapis.com/covid19-open-data/v2/main.csv
* Latest: https://storage.googleapis.com/covid19-open-data/v2/latest/main.csv
* Last 7 days: https://storage.googleapis.com/covid19-open-data/v2/7/main.csv
* Last 14 days: https://storage.googleapis.com/covid19-open-data/v2/14/main.csv
* Last 30 days: https://storage.googleapis.com/covid19-open-data/v2/30/main.csv

Note that the `latest` version contains the last non-null record for each key, whereas all others
contain the last `N` days of data (all of which could be null for some keys). All of the above
listed tables have a corresponding JSON version; simply replace `csv` with `json` in the link.

If you are trying to use this data alongside your own datasets, then you can use the [Index](#index)
table to get access to the ISO 3166 / NUTS / FIPS code, although administrative subdivisions are
not consistent among all reporting regions. For example, for the intra-country reporting, some EU
countries use NUTS2, others NUTS3 and many ISO 3166-2 codes.

You can find several examples in the [examples subfolder](examples) with code showcasing how to load
and analyze the data for several programming environments. If you want the short version, here are a
few snippets to get started.

### Google Colab
You can use Google Colab if you want to run your analysis without having to install anything in your
computer, simply go to this URL: https://colab.research.google.com/github/open-covid-19/data.

### Google Sheets
You can import the data directly into Google Sheets, as long as you stay within the size limits.
For instance, the following formula loads the latest epidemiology data into the current sheet:
```python
=IMPORTDATA("https://storage.googleapis.com/covid19-open-data/v2/latest/epidemiology.csv")
```
Note that Google Sheets has a size limitation, so only data from the `latest` subfolder can be
imported automatically. To work around that, simply download the file and import it via the File
menu.

### R
If you prefer R, then this is all you need to do to load the epidemiology data:
```R
data <- read.csv("https://storage.googleapis.com/covid19-open-data/v2/main.csv")
```

### Python
In Python, you need to have the package `pandas` installed to get started:
```python
import pandas
data = pandas.read_csv("https://storage.googleapis.com/covid19-open-data/v2/main.csv")
```

### jQuery
Loading the JSON file using jQuery can be done directly from the output folder,
this code snippet loads the epidemiology table into the `data` variable:
```javascript
$.getJSON("https://storage.googleapis.com/covid19-open-data/v2/epidemiology.json", data => { ... }
```

### Powershell
You can also use Powershell to get the latest data for a country directly from
the command line, for example to query the latest epidemiology data for Australia:
```powershell
Invoke-WebRequest 'https://storage.googleapis.com/covid19-open-data/v2/latest/epidemiology.csv' | ConvertFrom-Csv | `
    where key -eq 'AU' | select date,total_confirmed,total_deceased,total_recovered
```



## Understand the data

Make sure that you are using the URL [linked at the table above](#open-covid-19-dataset) and not the
raw GitHub file, the latter is subject to change at any moment in non-compatible ways, and due to
the configuration of GitHub's raw file server you may run into potential caching issues.

Missing values will be represented as nulls, whereas zeroes are used when a true value of zero is
reported.

### Main
Flat table with records from all other tables joined by `key` and `date`. See below for information
about all the different tables and columns. Tables not included in the main table are:
* [WorldBank](#worldbank): A subset of individual indicators are added as columns to other tables
  instead; for example, the [health](#health) table.
* [WorldPop](#worldpop): Age and sex demographics breakdowns are normalized and added to the
  [demographics](#demographics) table instead.
* [By Age](#by-age): Age breakdowns of epidemiology and hospitalization data are normalized and
  added to the by-age-normalized table instead (TABLE TO BE ADDED).
* [By Sex](#by-sex): Sex breakdowns of epidemiology and hospitalization data are normalized and
  added to the by-sex-normalized table instead (TABLE TO BE ADDED).

### Index
Non-temporal data related to countries and regions. It includes keys, codes and names for each
region, which is helpful for displaying purposes or when merging with other data:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **key** | `string` | Unique string identifying the region | US_CA_06001 |
| **wikidata** | `string` | Wikidata ID corresponding to this key | Q107146 |
| **datacommons** | `string` | DataCommons ID corresponding to this key | geoId/06001 |
| **country_code** | `string` | ISO 3166-1 alphanumeric 2-letter code of the country | US |
| **country_name** | `string` | American English name of the country, subject to change | United States of America |
| **subregion1_code** | `string` | (Optional) ISO 3166-2 or NUTS 2/3 code of the subregion | CA |
| **subregion1_name** | `string` | (Optional) American English name of the subregion, subject to change | California |
| **subregion2_code** | `string` | (Optional) FIPS code of the county (or local equivalent) | 06001 |
| **subregion2_name** | `string` | (Optional) American English name of the county (or local equivalent), subject to change | Alameda County |
| **3166-1-alpha-2** | `string` | ISO 3166-1 alphanumeric 2-letter code of the country | US |
| **3166-1-alpha-3** | `string` | ISO 3166-1 alphanumeric 3-letter code of the country | USA |
| **aggregation_level** | `integer` `[0-2]` | Level at which data is aggregated, i.e. country, state/province or county level | 2 |

### Demographics
Information related to the population demographics for each region:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **key** | `string` | Unique string identifying the region | KR |
| **population** | `integer` | Total count of humans | 51606633 |
| **population_male** | `integer` | Total count of males | 25846211 |
| **population_female** | `integer` | Total count of females | 25760422 |
| **rural_population** | `integer` | Population in a rural area | 9568386 |
| **urban_population** | `integer` | Population in an urban area | 42038247 |
| **largest_city_population** | `integer` | Population in the largest city of the region | 9963497 |
| **clustered_population** | `integer` | Population in urban agglomerations of more than 1 million | 25893097 |
| **population_density** | `double` `[persons per squared kilometer]` | Population per squared kilometer of land area | 529.3585 |
| **human_development_index** | `double` `[0-1]` | Composite index of life expectancy, education, and per capita income indicators | 0.903 |
| **population_age_`${lower}`_`${upper}`** | `integer` | Estimated population between the ages of `${lower}` and `${upper}`, both inclusive | 42038247 |

### Economy
Information related to the economic development for each region:

| Name | Name | Description | Example |
| ---- | ---- | ----------- | ------- |
| **key** | `string` | Unique string identifying the region | CN_HB |
| **gdp** | `integer` `[USD]` | Gross domestic product; monetary value of all finished goods and services | 24450604878 |
| **gdp_per_capita** | `integer` `[USD]` | Gross domestic product divided by total population | 1148 |
| **human_capital_index** | `double` `[0-1]` | Mobilization of the economic and professional potential of citizens | 0.765 |

### Epidemiology
Information related to the COVID-19 infections for each date-region pair:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **date** | `string` | ISO 8601 date (YYYY-MM-DD) of the datapoint | 2020-03-30 |
| **key** | `string` | Unique string identifying the region | CN_HB |
| **new_confirmed<sup>1</sup>** | `integer` | Count of new cases confirmed after positive test on this date | 34 |
| **new_deceased<sup>1</sup>** | `integer` | Count of new deaths from a positive COVID-19 case on this date | 2 |
| **new_recovered<sup>1</sup>** | `integer` | Count of new recoveries from a positive COVID-19 case on this date | 13 |
| **new_tested<sup>2</sup>** | `integer` | Count of new COVID-19 tests performed on this date | 13 |
| **total_confirmed<sup>3</sup>** | `integer` | Cumulative sum of cases confirmed after positive test to date | 6447 |
| **total_deceased<sup>3</sup>** | `integer` | Cumulative sum of deaths from a positive COVID-19 case to date | 133 |
| **total_tested<sup>2,3</sup>** | `integer` | Cumulative sum of COVID-19 tests performed to date | 133 |

<sup>1</sup>Values can be negative, typically indicating a correction or an adjustment in the way
they were measured. For example, a case might have been incorrectly flagged as recovered one date so
it will be subtracted from the following date.\
<sup>2</sup>When the reporting authority makes a distinction between PCR and antibody testing, only
PCR tests are reported here.\
<sup>3</sup>Total count will not always amount to the sum of daily counts, because many authorities
make changes to criteria for counting cases, but not always make adjustments to the data. There is
also potential missing data. All of that makes the total counts *drift* away from the sum of all
daily counts over time, which is why the cumulative values, if reported, are kept in a separate
column.

### Geography
Information related to the geography for each region:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **key** | `string` | Unique string identifying the region | CN_HB |
| **latitude** | `double` | Floating point representing the geographic coordinate | 30.9756 |
| **longitude** | `double` | Floating point representing the geographic coordinate | 112.2707 |
| **elevation** | `integer` `[meters]` | Elevation above the sea level | 875 |
| **area** | `integer` [squared kilometers] | Area encompassing this region | 3729 |
| **rural_area** | `integer` [squared kilometers] | Area encompassing rural land in this region | 3729 |
| **urban_area** | `integer` [squared kilometers] | Area encompassing urban land this region | 3729 |

### Health
Health related indicators for each region:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **key** | `string` | Unique string identifying the region | BN |
| **life_expectancy** | `double` `[years]` |Average years that an individual is expected to live | 75.722 |
| **smoking_prevalence** | `double` `[%]` | Percentage of smokers in population | 16.9 |
| **diabetes_prevalence** | `double` `[%]` | Percentage of persons with diabetes in population | 13.3 |
| **infant_mortality_rate** | `double` | Infant mortality rate (per 1,000 live births) | 9.8 |
| **adult_male_mortality_rate** | `double` | Mortality rate, adult, male (per 1,000 male adults) | 143.719 |
| **adult_female_mortality_rate** | `double` | Mortality rate, adult, female (per 1,000 male adults) | 98.803 |
| **pollution_mortality_rate** | `double` | Mortality rate attributed to household and ambient air pollution, age-standardized (per 100,000 population) | 13.3 |
| **comorbidity_mortality_rate** | `double` `[%]` | Mortality from cardiovascular disease, cancer, diabetes or cardiorespiratory disease between exact ages 30 and 70 | 16.6 |
| **hospital_beds** | `double` | Hospital beds (per 1,000 people) | 2.7 |
| **nurses** | `double` | Nurses and midwives (per 1,000 people) | 5.8974 |
| **physicians** | `double` | Physicians (per 1,000 people) | 1.609 |
| **health_expenditure** | `double` `[USD]` | Health expenditure per capita | 671.4115 |
| **out_of_pocket_health_expenditure** | `double` `[USD]` | Out-of-pocket health expenditure per capita | 34.756348 |

Note that the majority of the health indicators are only available at the country level.

### Hospitalizations
Information related to patients of COVID-19 and hospitals:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **key** | `string` | Unique string identifying the region | CN_HB |
| **new_hospitalized\*** | `integer` | Count of new cases hospitalized after positive test on this date | 34 |
| **new_intensive_care\*** | `integer` | Count of new cases admitted into ICU after a positive COVID-19 test on this date | 2 |
| **new_ventilator\*** | `integer` | Count of new COVID-19 positive cases which require a ventilator on this date | 13 |
| **total_hospitalized\*\*** | `integer` | Cumulative sum of cases hospitalized after positive test to date | 6447 |
| **total_intensive_care\*\*** | `integer` | Cumulative sum of cases admitted into ICU after a positive COVID-19 test to date | 133 |
| **total_ventilator\*\*** | `integer` | Cumulative sum of COVID-19 positive cases which require a ventilator to date | 133 |
| **current_hospitalized\*\*** | `integer` | Count of current (active) cases hospitalized after positive test to date | 34 |
| **current_intensive_care\*\*** | `integer` | Count of current (active) cases admitted into ICU after a positive COVID-19 test to date | 2 |
| **current_ventilator\*\*** | `integer` | Count of current (active) COVID-19 positive cases which require a ventilator to date | 13 |

\*Values can be negative, typically indicating a correction or an adjustment in the way they were
measured. For example, a case might have been incorrectly flagged as recovered one date so it will
be subtracted from the following date.

\*\*Total count will not always amount to the sum of daily counts, because many authorities make
changes to criteria for counting cases, but not always make adjustments to the data. There is also
potential missing data. All of that makes the total counts *drift* away from the sum of all daily
counts over time, which is why the cumulative values, if reported, are kept in a separate column.

### Mobility
[Google's][17] Mobility Reports are presented in CSV form as
[mobility.csv](https://storage.googleapis.com/covid19-open-data/v2/mobility.csv) with the following columns:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **date** | `string` | ISO 8601 date (YYYY-MM-DD) of the datapoint | 2020-03-30 |
| **key** | `string` | Unique string identifying the region | US_CA |
| **mobility_transit_stations** | `double` `[%]` |  Percentage change in visits to transit station locations compared to baseline | -15 |
| **mobility_retail_and_recreation** | `double` `[%]` |  Percentage change in visits to retail and recreation locations compared to baseline | -15 |
| **mobility_grocery_and_pharmacy** | `double` `[%]` |  Percentage change in visits to grocery and pharmacy locations compared to baseline | -15 |
| **mobility_parks** | `double` `[%]` |  Percentage change in visits to park locations compared to baseline | -15 |
| **mobility_residential** | `double` `[%]` |  Percentage change in visits to residential locations compared to baseline | -15 |
| **mobility_workplaces** | `double` `[%]` |  Percentage change in visits to workplace locations compared to baseline | -15 |

### Oxford Government Response
Summary of a government's response to the events, including a *stringency index*, collected from
[University of Oxford][18]:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **date** | `string` | ISO 8601 date (YYYY-MM-DD) of the datapoint | 2020-03-30 |
| **key** | `string` | Unique string identifying the region | US_CA |
| **school_closing** | `integer` `[0-3]` | Schools are closed | 2 |
| **workplace_closing** | `integer` `[0-3]` | Workplaces are closed | 2 |
| **cancel_public_events** | `integer` `[0-3]` | Public events have been cancelled | 2 |
| **restrictions_on_gatherings** | `integer` `[0-3]` | Gatherings of non-household members are restricted | 2 |
| **public_transport_closing** | `integer` `[0-3]` | Public transport is not operational | 0 |
| **stay_at_home_requirements** | `integer` `[0-3]` | Self-quarantine at home is mandated for everyone | 0 |
| **restrictions_on_internal_movement** | `integer` `[0-3]` | Travel within country is restricted | 1 |
| **international_travel_controls** | `integer` `[0-3]` | International travel is restricted | 3 |
| **income_support** | `integer` `[USD]` | Value of fiscal stimuli, including spending or tax cuts | 20449287023 |
| **debt_relief** | `integer` `[0-3]` | Debt/contract relief for households | 0 |
| **fiscal_measures** | `integer` `[USD]` | Value of fiscal stimuli, including spending or tax cuts | 20449287023 |
| **international_support** | `integer` `[USD]` | Giving international support to other countries | 274000000 |
| **public_information_campaigns** | `integer` `[0-2]` | Government has launched public information campaigns | 1 |
| **testing_policy** | `integer` `[0-3]` | Country-wide COVID-19 testing policy | 1 |
| **contact_tracing** | `integer` `[0-2]` | Country-wide contact tracing policy | 1 |
| **emergency_investment_in_healthcare** | `integer` `[USD]` | Emergency funding allocated to healthcare | 500000 |
| **investment_in_vaccines** | `integer` `[USD]` | Emergency funding allocated to vaccine research | 100000 |
| **stringency_index** | `double` `[0-100]` | Overall stringency index | 71.43 |

For more information about each field and how the overall stringency index is
computed, see the [Oxford COVID-19 government response tracker][18].

### Weather
Daily weather information from nearest station reported by NOAA:

| Name | Type | Description | Example |
| ---- | ---- | ----------- | ------- |
| **date** | `string` | ISO 8601 date (YYYY-MM-DD) of the datapoint | 2020-03-30 |
| **key** | `string` | Unique string identifying the region | US_CA |

CHANGELOG.md
# Changelog

### Open COVID-19 V2
Preview of V2 finalized:
* Audited all data sources
* Improved reliability of data
* Implemented new data ingestion pipeline
* Changed column names to snake_case to improve readability
* Added capability for 3 levels of region hierarchy (country, state/province and county)
* Broke down tables into: metadata, epidemiology, demographics, geography, economy, stringency
* Matched every region with a wikidata ID
* Added tested and recovered data to epidemiology table
* Added county-level data for CO, GB, NL and US
* Now tracking both new and cumulative cases


src/publish.py
#!/usr/bin/env python
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import json
import shutil
import datetime
import cProfile
from pstats import Stats
from pathlib import Path
from functools import partial
from argparse import ArgumentParser

from pandas import DataFrame, date_range

from lib.concurrent import thread_map
from lib.io import read_file, export_csv, pbar
from lib.utils import SRC, drop_na_records


def subset_last_days(output_folder: Path, days: int) -> None:
    """ Outputs last N days of data """
    n_days_folder = output_folder / str(days)
    n_days_folder.mkdir(exist_ok=True)
    for csv_file in (output_folder).glob("*.csv"):
        table = read_file(csv_file, low_memory=False)

        # Degenerate case: this table has no date
        if not "date" in table.columns or len(table.date.dropna()) == 0:
            export_csv(table, n_days_folder / csv_file.name)
        else:
            last_date = datetime.date.fromisoformat(max(table.date))
            # Since APAC is almost always +1 days ahead, increase the window by 1
            first_date = last_date - datetime.timedelta(days=days + 1)
            export_csv(table[table.date >= first_date.isoformat()], n_days_folder / csv_file.name)


def subset_latest(output_folder: Path, csv_file: Path) -> DataFrame:
    """ Outputs latest data for each key """
    latest_folder = output_folder / "latest"
    latest_folder.mkdir(exist_ok=True)
    table = read_file(csv_file, low_memory=False)

    # Degenerate case: this table has no date
    if not "date" in table.columns or len(table.date.dropna()) == 0:
        return export_csv(table, latest_folder / csv_file.name)
    else:
        non_null_columns = [col for col in table.columns if not col in ("key", "date")]
        table = table.dropna(subset=non_null_columns, how="all")
        table = table.sort_values("date").groupby(["key"]).tail(1).reset_index()
        export_csv(table, latest_folder / csv_file.name)


def subset_grouped_key(table_indexed: DataFrame, output_folder: Path, key: str) -> None:
    """ Outputs a subset of the table with only records with the given key """
    key_folder = output_folder / key
    key_folder.mkdir(exist_ok=True)
    export_csv(table_indexed.loc[key:key].reset_index(), key_folder / "main.csv")


def export_json_without_index(csv_file: Path) -> None:
    table = read_file(csv_file, low_memory=False)
    json_path = str(csv_file).replace("csv", "json")
    json_dict = json.loads(table.to_json(orient="split"))
    del json_dict["index"]
    with open(json_path, "w") as fd:
        json.dump(json_dict, fd)


def table_cross_product(table1: DataFrame, table2: DataFrame) -> DataFrame:
    tmp_col_name = "_tmp_cross_product_column"
    table1[tmp_col_name] = 1
    table2[tmp_col_name] = 1
    return table1.merge(table2, on=[tmp_col_name]).drop(columns=[tmp_col_name])


def main(output_folder: Path, tables_folder: Path, show_progress: bool = True) -> None:
    """
    This script takes the processed outputs located in `tables_folder` and publishes them into the
    output folder by performing the following operations:

        1. Copy all the tables as-is from `tables_folder` to `output_folder`
        2. Produce a main table, created by iteratively performing left outer joins on all other
           tables (with a few exceptions)
        3. Create different slices of data, such as the latest known record for each region, files
           for the last N days of data, files for each individual region
    """
    # TODO: respect disable progress flag
    disable_progress = not show_progress

    # Wipe the output folder first
    for item in output_folder.glob("*"):
        if item.name.startswith("."):
            continue
        if item.is_file():
            item.unlink()
        else:
            shutil.rmtree(item)

    # Create the folder which will be published using a stable schema
    v2_folder = output_folder / "v2"
    v2_folder.mkdir(exist_ok=True, parents=True)

    # Copy all output files to the V2 folder
    for output_file in pbar([*tables_folder.glob("*.csv")], desc="Copy tables"):
        shutil.copy(output_file, v2_folder / output_file.name)

    # Merge all output files into a single table
    main_table = read_file(v2_folder / "index.csv")

    # Add a date to each region from index to allow iterative left joins
    max_date = (datetime.datetime.now() + datetime.timedelta(days=7)).date().isoformat()
    date_list = [date.date().isoformat() for date in date_range("2020-01-01", max_date)]
    date_table = DataFrame(date_list, columns=["date"], dtype=str)
    main_table = table_cross_product(main_table, date_table)

    # Some tables are not included into the main table
    exclude_from_main_table = (
        "main.csv",
        "index.csv",
        "worldbank.csv",
        "worldpop.csv",
        "by-age.csv",
        "by-sex.csv",
    )

    non_dated_columns = set(main_table.columns)
    for output_file in pbar([*v2_folder.glob("*.csv")], desc="Main table"):
        if output_file.name not in exclude_from_main_table:
            # Load the table and perform left outer join
            table = read_file(output_file, low_memory=False)
            main_table = main_table.merge(table, how="left")
            # Keep track of columns which are not indexed by date
            if not "date" in table.columns:
                non_dated_columns = non_dated_columns | set(table.columns)

    # There can only be one record per <key, date> pair
    main_table = main_table.groupby(["key", "date"]).first().reset_index()

    # Drop rows with null date or without a single dated record
    main_table = drop_na_records(main_table.dropna(subset=["date"]), non_dated_columns)
    export_csv(main_table, v2_folder / "main.csv")

    # Create subsets with the last 30, 14 and 7 days of data
    map_func = partial(subset_last_days, v2_folder)
    for _ in thread_map(map_func, (30, 14, 7), desc="Last N days subsets"):
        pass

    # Create a subset with the latest known day of data for each key
    map_func = partial(subset_latest, v2_folder)
    for _ in thread_map(map_func, [*(v2_folder).glob("*.csv")], desc="Latest subset"):
        pass

    # Create subsets with each known key
    main_indexed = main_table.set_index("key")
    map_func = partial(subset_grouped_key, main_indexed, v2_folder)
    for _ in thread_map(map_func, main_indexed.index.unique(), desc="Grouped key subsets"):
        pass

    # Convert all CSV files to JSON using values format
    map_func = export_json_without_index
    for _ in thread_map(map_func, [*(v2_folder).glob("**/*.csv")], desc="JSON conversion"):
        pass


if __name__ == "__main__":

    # Process command-line arguments
    output_root = SRC / ".." / "output"
    argparser = ArgumentParser()
    argparser.add_argument("--profile", action="store_true")
    argparser.add_argument("--no-progress", action="store_true")
    argparser.add_argument("--tables-folder", type=str, default=str(output_root / "tables"))
    argparser.add_argument("--output-folder", type=str, default=str(output_root / "public"))
    args = argparser.parse_args()

    if args.profile:
        profiler = cProfile.Profile()
        profiler.enable()

    main(Path(args.output_folder), Path(args.tables_folder), show_progress=not args.no_progress)

    if args.profile:
        stats = Stats(profiler)
        stats.strip_dirs()
        stats.sort_stats("cumtime")
        stats.print_stats(20)
